sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala

```diff
@@ -95,6 +112,50 @@ object ExtractValue {
 
 trait ExtractValue extends Expression
 
+object ExtractNestedArray {
+
+  type ReturnType = Option[(DataType, Boolean, Seq[Boolean])]
+
+  def unapply(dataType: DataType): ReturnType = {
+     extractArrayType(dataType)
+  }
+
+  def extractArrayType(dataType: DataType): ReturnType = {
+    dataType match {
+      case ArrayType(dt, containsNull) =>
+        extractArrayType(dt) match {
+          case Some((d, cn, seq)) => Some(d, cn, containsNull +: seq)
+          case None => Some(dt, containsNull, Seq.empty[Boolean])
+        }
+      case _ => None
+    }
+  }
+}
+
+/**
+ * Extract GetArrayStructField from Expression
+ */
+object ExtractGetArrayStructField {
+
+  type ReturnType = Option[(Expression, Int)]
+
+  def unapply(expr: Expression): ReturnType = {
+    extractArrayStruct(expr)
+  }
+
+  def extractArrayStruct(expr: Expression): ReturnType = {
+    expr match {
+      case gas @ GetArrayStructFields(child, _, _, _, _) =>
```

Comment:
Let's avoid arguments matching. This is actually an anti pattern - https://github.com/databricks/scala-style-guide#pattern-matching